forem/forem
# Close hamburger menu on click

## What type of PR is this? (check all applicable)

## Description
**I need some help writing tests and finding the best approach to this bug.**

I'll explain the bug here and in PR code comments below.

There are a few layers to unpack here for the full context, so bear with me 😅.

The current issue is that when a user is _signed out_, the hamburger menu does not close when you click on a link. Upon investigation, there are a couple of things at play.

First, the reason it works _signed in_ is actually a happy accident. To toggle the hamburger menu we use the `hamburger-open` class on the `body` element:

https://github.com/forem/forem/blob/09abd1af81f65a32b0d3f9e2c370bff65eac2c2f/app/javascript/topNavigation/utilities.js#L105

The problem is, when a user is _signed in_ we completely overwrite the classes on the `body` to setup configs:
https://github.com/forem/forem/blob/09abd1af81f65a32b0d3f9e2c370bff65eac2c2f/app/views/layouts/_user_config.html.erb#L7

This logic overwrites the `hamburger-open` class which makes it seem like it's working when you're signed in. That's actually unintentional. See more below in PR review code comments.

Back to addressing this when the user is _signed out_. Another piece of the problem is how we attach the `click` events to the hamburger menu. The links themselves in the hamburger menu did not have the necessary `js-hamburger-trigger` classes on them. That's what we use to toggle the menu.

Step 1 - grab all the `js-hamburger-trigger` elements on the page:
https://github.com/forem/forem/blob/09abd1af81f65a32b0d3f9e2c370bff65eac2c2f/app/javascript/packs/base.js#L17

Step 2 - call `initializeMobileMenu` on these elements
https://github.com/forem/forem/blob/09abd1af81f65a32b0d3f9e2c370bff65eac2c2f/app/javascript/packs/base.js#L33

Step 3 - add the `click` events to each of these elements
https://github.com/forem/forem/blob/09abd1af81f65a32b0d3f9e2c370bff65eac2c2f/app/javascript/topNavigation/utilities.js#L144-L147

To fix this, I added the `js-hamburger-trigger` classes to each link in the hamburger menu so that it correctly triggers the menu toggle on click.

## Related Tickets & Documents
closes https://github.com/forem/forem/issues/12645

## QA Instructions, Screenshots, Recordings
Fire up the app locally and follow the following steps twice - once **signed in** and again **signed out**.

1. Shrink your screen size small enough to force the hamburger menu in the top left corner to show up.
2. Click on each link in the hamburger menu.
3. Ensure the link opens and the hamburger menu closes for each item 🎉 .

![Screen Shot 2021-02-17 at 4 06 44 PM](https://user-images.githubusercontent.com/15987080/108268295-4be0d600-713a-11eb-9de5-0614acf9bdc1.png)

## Added tests?
- [x] I need help with writing tests. **How would you recommend writing tests for this one?**

## [Forem core team only] How will this change be communicated?
- [x] This change does not need to be communicated, and this is why not: this a bug fix introducing the expected behavior.

![tom_brady_post_game_gif](https://media.giphy.com/media/1GcgoCwd2VzO9M2l5T/giphy.gif)
